File: dotnet-soa/ProductService/Domain/Models/Product.cs
namespace ProductService.Domain.Models;

/// <summary>
/// Product domain model representing the core business entity.
/// 
/// SOA Principle: Service contracts are defined through domain models
/// that represent the service's bounded context.
/// </summary>
public record Product
{
    public Guid Id { get; init; } = Guid.NewGuid();
    public required string Name { get; init; }
    public string? Description { get; init; }
    public required decimal Price { get; init; }
    public string? Category { get; init; }
    public DateTime CreatedAt { get; init; } = DateTime.UtcNow;
    public DateTime UpdatedAt { get; init; } = DateTime.UtcNow;
}

File: nodejs-soa/product-service/src/domain/model/Product.ts
// =============================================================================
// Product Domain Entity
// =============================================================================
// The domain model represents the core business concept.
// In SOA, domain entities are internal to the service and never exposed directly.
// External communication uses DTOs (Data Transfer Objects).
// =============================================================================

/**
 * Product domain entity - represents a product in the catalog.
 * This is the core business object that the Product Service manages.
 */
export interface Product {
  id: string;
  name: string;
  description: string | null;
  price: number;
  category: string | null;
  sku: string;
  active: boolean;
  createdAt: Date;
  updatedAt: Date;
}

/**
 * Data required to create a new product.
 * Separating creation data from the full entity follows the
 * Command Query Responsibility Segregation (CQRS) principle.
 */
export interface CreateProductData {
  name: string;
  description?: string;
  price: number;
  category?: string;
  sku: string;
}

/**
 * Data for updating an existing product.
 * All fields are optional - only provided fields will be updated.
 */
export interface UpdateProductData {
  name?: string;
  description?: string;
  price?: number;
  category?: string;
  active?: boolean;
}
